akka-stream-tests/src/test/scala/akka/stream/scaladsl/FoldWithinSpec.scala

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (C) 2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.scaladsl
+
+import akka.stream.impl.fusing.FoldWithin
+import akka.stream.testkit.StreamSpec
+
+import scala.concurrent.{Await, Future}
+import scala.concurrent.duration._
+
+class FoldWithinSpec extends StreamSpec {
+
+  "split aggregator by size" in {
+    val stream    = collection.immutable.Seq(1, 2, 3, 4, 5, 6, 7)
+    val groupSize = 3
+    val result = Source(stream)
+      .via(
+        new FoldWithin[Int, Seq[Int], Seq[Int]](
+          seed = i => Seq(i),
+          aggregate = (seq, i) => seq :+ i,
+          emitReady = seq => seq.size >= groupSize,
+          harvest = seq => seq
+        )
+      )
+      .runWith(Sink.collection)
+
+    assert(
+      Await.result(result, 10.seconds) == stream.grouped(groupSize).toSeq
+    )
+  }
+
+  "split aggregator by size and harvest" in {
+    val stream    = collection.immutable.Seq(1, 2, 3, 4, 5, 6, 7)
+    val groupSize = 3
+    val result = Source(stream)
+      .via(
+        new FoldWithin[Int, Seq[Int], Seq[Int]](
+          seed = i => Seq(i),
+          aggregate = (seq, i) => seq :+ i,
+          emitReady = seq => seq.size >= groupSize,
+          harvest = seq => seq :+ -1 // append -1 to output to demonstrate harvest
+        )
+      )
+      .runWith(Sink.collection)
+
+    assert(
+      Await.result(result, 10.seconds) == stream.grouped(groupSize).toSeq.map(seq => seq :+ -1)
+    )
+  }
+
+  "split aggregator by custom weight condition" in {
+    val stream = collection.immutable.Seq(1, 2, 3, 4, 5, 6, 7)
+    val weight = 10
+    // use the value as weight and aggregate
+    val result = Source(stream)
+      .via(
+        new FoldWithin[Int, (Seq[Int], Int), Seq[Int]](
+          seed = i => (Seq(i), i),
+          aggregate = (seqAndWeight, i) => (seqAndWeight._1 :+ i, seqAndWeight._2 + i),
+          emitReady = seqAndWeight => seqAndWeight._2 >= weight,
+          harvest = seqAndWeight => seqAndWeight._1
+        )
+      )
+      .runWith(Sink.collection)
+
+    assert(
+      Await.result(result, 10.seconds) ==
+        Seq(Seq(1, 2, 3, 4), Seq(5, 6), Seq(7))
+    )
+  }
+
+  "split aggregator by gap for slow upstream" in {
+    val stream = collection.immutable.Seq(
+      (1, 10),
+      (2, 10),
+      (3, 100),
+      (4, 10),
+      (5, 100),
+      (6, 10),
+      (7, 10)
+    )
+
+    val maxGap = 20.millis
+
+    val result = Source(stream)
+      .mapAsync(1) {
+        case (i, gap) =>
+          Future {
+            Thread.sleep(gap)
```

Comment:
There is a ManualTime using this class to control the system scheduler.
It looks like the right thing based on the doc, but I cannot find existing example tests using this.
I'll investigate. Please let me know if there are good examples I can use as template.

Reply:
No usage except the typed teskit example `ManualTimerExampleSpec` so not sure if there are some specific caveats for stream usage.

The ´ManualTimer` is in the Typed Actor TestKit though, which I don't think the stream tests has a dependency on, but the actual fake clock is `akka.testkit.ExplicitlyTriggeredScheduler` which is in the classic Actor TestKit so that should be readily available without changing any dependencies (but with a little less convenience compared to `ManualTimer`). Since it is a test module I think it would be fine to add a dependency on the typed testkit as well if faking time works out nicely.

Reply:
Even the scheduler triggered event can follow the faked time this way, the implementation depends on comparing the actual `System.nanoTime`. This approach won't work unless the system clock can also be faked transparently by some supported method built-in with this testing scheduler. Is there something like that? Does this sound right?

Reply:
Ah, ofc, I didn't think of that. So you'd have to hide the nanoTime behind something that can be replaced in the test, I'm not entirely sure that is worth it, wdyt @patriknw any better ideas?

Reply:
this line of thinking brought back the `scheduleOnce` approach again. It does look like it has lots of overhead, but that is a more accurate interval and make the rest of the logic completely event driven independent of the actual system time. Would be nice if we can just switch to that in testing mode only. I think `scheduleOnce` approach can work with the fake scheduler. The only tricky part is to just "believe" the fixed delay timer is functionally equivalent (event not exactly the same) without testing it. Not sure that even makes sense from testing perspectives.

Reply:
There is also potential benefits to use `scheduleOnce` as a configurable alternative. It has wore worst case and better best case comparing to fixed delay scheduler. I can be complimentary, and easy to write end-to-end functional test case.

Reply:
it's pretty straightforward to expose system time in the `Scheduler` trait and let `akka.testkit.ExplicitlyTriggeredScheduler` override it. GraphStage has access to the scheduler through the materializer, so it's easy to setup testing as well

Reply:
> can we find another way to test this than timing sensitive sleeps? we have a few other tests that are flaky because of this and tracking it down and hardening it afterwards is much work for the maintainers of Akka
> 
> CI environments can be very non-deterministic for timing sensitive things. Expect random pauses up to a few seconds.

@patriknw Just want to confirm you are looking for a different approach to do timing test, ideally based on the simulated time. I found many tests using the method `expectNext(duration, ) ` which is based on real time with `taggedAs TimingTest`. Is this the problematic tests you are referring to?

Reply:
I know it's very difficult to test time related things. It's also a matter of that one don't want to fake time too much because then in the end we are not testing the real thing.

I'm skeptical to that this will work with sleeps if the precision is like this with 10s or 100s of milliseconds. You must think in terms of seconds when running in a flaky CI environment.

I'm fine with using real time and have tests marked as `TimingTest` but it should still work in CI environment.

Reply:
I think I found a way to use entirely simulated time for test with the `ExplicitlyTriggeredScheduler`. Then I saw these existing TimingTest that seems to be based on realtime. https://github.com/akka/akka/blob/main/akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowGroupedWithinSpec.scala#L43
I’m not very familiar with those existing tests but they do seem to have limitations. I assume the reason of the tagging is to make them sequentially run. If you can confirm the existing tests are “no good”, I can feel more confident to go down my path. Otherwise I don’t feel justified to add something new if the existing testing framework already support the same functionality.